internal/pkg/service/common/servicectx/servicectx.go

```diff
@@ -79,50 +100,59 @@ func New(opts ...Option) (*Process, error) {
 		c.uniqueID = fmt.Sprintf(`%s-%05d`, hostname, pid)
 	}
 
-	// Create channel used by both the signal handler and service goroutines
-	// to notify the main goroutine when to stop the server.
-	errCh := make(chan error, 1)
-
-	// Create channel used to pass the shutdown context to OnShutdownFn callbacks.
-	shutdownCtxCh := make(chan context.Context, 1)
-
-	proc := &Process{
-		logger:        c.logger,
-		wg:            &sync.WaitGroup{},
-		errCh:         errCh,
-		uniqueID:      c.uniqueID,
-		lock:          &sync.Mutex{},
-		shutdownCtxCh: shutdownCtxCh,
+	v := &Process{
+		uniqueID:    c.uniqueID,
+		logger:      c.logger,
+		wg:          &sync.WaitGroup{},
+		done:        make(chan struct{}),
+		lock:        &sync.Mutex{},
+		terminating: make(chan struct{}),
+		shutdown:    make(chan context.Context, 1),
 	}
 
-	// Register onShutdown operation
-	proc.Add(func(ctx context.Context, _ ShutdownFn) {
-		<-ctx.Done()
-		shutdownCtx := <-proc.shutdownCtxCh
+	// Execute OnShutdown callbacks and wait for them
+	v.wg.Add(1)
+	go func() {
+		defer v.wg.Done()
+
+		// Wait for shutdown, see Shutdown function
+		shutdownCtx := <-v.shutdown
 
 		// Iterate callbacks in reverse order, LIFO, see the OnShutdown method
-		for i := len(proc.onShutdown) - 1; i >= 0; i-- {
-			proc.onShutdown[i](shutdownCtx)
+		for i := len(v.onShutdown) - 1; i >= 0; i-- {
+			v.onShutdown[i](shutdownCtx)
 		}
-	})
+	}()
 
-	// Setup interrupt handler,
-	// so SIGINT and SIGTERM signals cause the services to stop gracefully.
+	// Setup interrupt handler, so SIGINT and SIGTERM signals trigger shutdown.
 	go func() {
 		sigCh := make(chan os.Signal, 1)
 		signal.Notify(sigCh, syscall.SIGINT, syscall.SIGTERM)
 
 		select {
-		case <-ctx.Done():
-			// Shutdown was triggered by another way
 		case sig := <-sigCh:
-			// Shutdown signal
-			proc.Shutdown(ctx, errors.Errorf("%s", sig))
+			// Trigger shutdown on signal
+			v.Shutdown(context.Background(), errors.Errorf("%s", sig))
+		case <-v.terminating:
+			// The process was shut down by another trigger, stop goroutine
+			return
 		}
 	}()
 
-	proc.logger.Infof(`process unique id "%s"`, proc.UniqueID())
-	return proc, nil
+	// Finalizer
```

Comment:
The goroutine logs `exited` message and unblock `WaitForShutdown` method calls.

In the previous version, the code was part of the first `WaitForShutdown` call.
Here it is more clear, and `WaitForShutdown` method is simple.